3p/ampanalytics-lib.js

```diff
@@ -0,0 +1,264 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import './polyfills';
+import {tryParseJson} from '../src/json';
+import {dev, user, initLogConstructor, setReportError} from '../src/log';
+import {IFRAME_TRANSPORT_EVENTS_TYPE} from '../src/iframe-transport-common';
+import {getData} from '../src/event-helper';
+
+initLogConstructor();
+// TODO(alanorozco): Refactor src/error.reportError so it does not contain big
+// transitive dependencies and can be included here.
+setReportError(() => {});
+
+/** @private @const {string} */
+const TAG_ = 'ampanalytics-lib';
+
+/**
+ * Receives event messages bound for this cross-domain iframe, from all
+ * creatives
+ */
+export class EventRouter {
+
+  /** @param {!Window} win */
+  constructor(win) {
+    /** @private {!Window} */
+    this.win_ = win;
+
+    /** @const {string} */
+    this.sentinel_ = user().assertString(
+        tryParseJson(this.win_.name)['sentinel'],
+        'Invalid/missing sentinel on iframe name attribute' + this.win_.name);
+    if (!this.sentinel_) {
+      return;
+    }
+
+    /**
+     * Multiple creatives on a page may wish to use the same type of
+     * amp-analytics tag. This object provides a mapping between the
+     * IDs which identify which amp-analytics tag a message is to/from,
+     * with each ID's corresponding CreativeEventRouter,
+     * which is an object that handles messages to/from a particular creative.
+     * @private {!Object<string, !CreativeEventRouter>}
+     */
+    this.creativeEventRouters_ = {};
+
+    this.win_.addEventListener('message', event => {
+      const messageContainer = this.extractMessage_(event);
+      if (this.sentinel_ != messageContainer['sentinel']) {
+        return;
+      }
+      user().assert(messageContainer['type'],
+          'Received message with missing type in ' + this.win_.location.href);
+      user().assert(messageContainer['events'],
+          'Received empty message in ' + this.win_.location.href);
+      user().assert(
+          messageContainer['type'] == IFRAME_TRANSPORT_EVENTS_TYPE,
+          'Received unrecognized message type ' + messageContainer['type'] +
+          ' in ' + this.win_.location.href);
+      this.processEventsMessage_(
+          /**
+           * @type {!Array<../src/iframe-transport-common.IframeTransportEvent>}
+           */
+          (messageContainer['events']));
+    }, false);
+
+    this.subscribeTo(IFRAME_TRANSPORT_EVENTS_TYPE);
+  }
+
+  /**
+   * Sends a message to the parent frame, requesting to subscribe to a
+   * particular message type
+   * @param messageType The type of message to subscribe to
+   * @VisibleForTesting
+   */
+  subscribeTo(messageType) {
```

Comment:
sorry if i didn't make it clear. 
SubscriptionApi and IframeMessagingClient is a pair. the former is installed in host page, the latter is for the iframe client.

that's how it works for our 3P ads iframe